### ISSUE TITLE
throw if detekt cannot find any files, i.e input was most likely configured wrong

### ISSUE THREAD

**Issue by robstoll** [detekt/detekt#3344]
## Expected Behavior
Detekt task exits at least with a non 0 code - preferably with a message that `input` was configured wrong

## Current Behavior
Detekt is happy

## Context
I had a project which was a multi-project build. I have configured detekt as follows: 
``` 
detekt{
 input = files(subprojects*.collect { it.projectDir })
}
``` 
At some point I simplified the structure and converted it to a single project. Detekt continued to run but did not analyse anything and I have not detekted (:wink:) that for a long time

**Comment by BraisGabin** [comment]
I can't reproduce your issue. Which version of detekt are you using? Could you provide a sample project to demostrate this?

**Comment by robstoll** [comment]
The latest one I reckon, you can try out https://github.com/robstoll/niok/commit/c4917338abc38bfe0d008b40cba8afa02380d558
With the next commit I fixed the `input` again and I had to fix a few issues detekt detected:
https://github.com/robstoll/niok/commit/fb376a43c86b571c98896e78389e491ac4fbaf8e

**Comment by schalkms** [comment]
How should a computer program detect whether such a configuration was on purpose or not. Is entirely possible that projects don't contain Kotlin but rather Java, Scala or source code of other JVM languages.
Detekt reports in all kinds of flavours show how many files have been analyzed.
I'm thinking about the following cases, where detekt is frequently used.

- Template projects that as the name suggests contain no Kotlin sources.
- Not every single project in multi-projects contains Kotlin sources.

> Detekt task exits at least with a non 0 code

Why should it exit with a different code? Which other static source code analyzers yield an error due to the set of input source files being empty?

**Comment by robstoll** [comment]
> Is entirely possible that projects don't contain Kotlin but rather Java, Scala or source code of other JVM languages.

I don't get why you think I am talking about Kotlin only here. is `detekt { input =...}` only for Kotlin?
Anyway, I am not talking about any specific language, the use case I am talking about here is that I have messed up `input` so that detekt has basically 0 files to analyse. That's almost always an error and not on purpose. Otherwise there is no point in setting up detekt at all IMO. There is also not a need for a flag such as `allow 0 files` IMO, one implement a flag on its on and not apply the detekt plugin in such cases

> Why should it exit with a different code? 

I guess you are in a sarcastic mood ifor something and can answer yourself why a program which has an erroneous setup exits with something different than 0 :wink:

**Comment by schalkms** [comment]
> I don't get why you think I am talking about Kotlin only here. is detekt { input =...} only for Kotlin?

Suppose `input = mydir`:
How should detekt know whether mydir contains no Kotlin source files on purpose or not?
It's entirely possible that a directory without Kotlin code at the given point in time is configured on purpose.

> input so that detekt has basically 0 files to analyse. 

This can be seen in the built-in reports of detekt. Why should detekt terminate because of that?
By the way, you can implement this behavior with detekt's custom report feature. If there is actually 0 analyzed code, you could throw an exception.

> I guess you are in a sarcastic mood or something and can answer yourself why a program which has an erroneous setup exits with something different than 0 😉

Excuse me, this wasn't meant sarcastic. It was a serious question.
Which other static source code analyzers yield an error and terminate due to the set of input source files being empty?

**Comment by robstoll** [comment]
> It's entirely possible that a directory without Kotlin code at the given point in time is configured on purpose.

I agree that the directory can be empty at the point of configuration but when running detekt then, IMO, detekt should fail if it does not need to analyse anything at all. I don't know detekt well enough, maybe my assumption that `input =` can only be specified once leads to the confusion here; or in other words, maybe I should change the title of the issue to `throw if detekt does not need to analyse any files` (again, I am not talking about Kotlin files exclusively, I am talking about 0 files of what any kind).

> Which other static source code analyzers yield an error and terminate due to the set of input source files being empty?

Don't know any on top of my head but I think it makes sense to error on misconfiguration instead of happily exit with 0 which in turn means the build/CI will not fail, giving the dev creating e.g. a PR as well as the reviewer a false view of the actual state (i.e. that the source code might be full of violations.) 

I know a test-runner which exits with 0 if it cannot find any tests to execute and IMO this is the same use case. junit5 fails if `--fail-if-no-tests` is specified

**Comment by cortinico** [comment]
> I know a test-runner which exits with 2 if it cannot find any tests to execute and IMO this is the same use case. junit5 fails if `--fail-if-no-tests` is specified

Technically we could add a flag in the config file to achieve this. I'm unsure of the usefulness of this flag as it seems more an edge case to me 🤔 

Also JUnit 5, as you mentioned, is returning a success if there are no tests to run (as they have the aformentioned flag).

Moreover the `detekt` task is a `SourceTask`. I don't know your exact detekt configuration, but you can probably add a `doLast` that will fail the task if `source` is empty or not.

**Comment by robstoll** [comment]
First of all, feel free to close this issue, it's an idea, but I can live without it (can add an own check). 
IMO a static analysis tool which tries to detect problems should also detect a wrong setup which is a problem as well and IMO a big problem because it covers up all real problems sitting there in the code. IMO this check should not be behind a flag but better behind a flag than nothing.

> I don't know your exact detekt configuration

This was the erroneous configuration:
https://github.com/robstoll/niok/commit/fb376a43c86b571c98896e78389e491ac4fbaf8e#diff-49a96e7eea8a94af862798a45174e6ac43eb4f8b4bd40759b5da63ba31ec3ef7L71

As you can see, I have mis-configured `input` and pointed it to an empty list. The project had once subprojects where this configuration was correct but not anymore and I forgot to change it when I refactored the project to a single project. Since detekt was always green, I did not detect it for a long time.

**Comment by cortinico** [comment]
I've done a bit more research on this front just to understand what was happening.

> but you can probably add a doLast that will fail the task if source is empty or not.

This is the check you can add to your `build.gradle` if you want your build to fail once the input is empty.

```groovy
 gradle.taskGraph.afterTask {
     if(it.state.noSource && it.path == ":detekt"){
         throw new StopExecutionException("Detekt has an empty input")
     }
 }
```

What happened in your case was that you provided an empty input for detekt. The `detekt` task is a `SourceTask` that exists with the `NO-SOURCE` status if the input is empty. Technically our code not ever runs, and Gradle just realise that the task has no input so it can be skipped.

This is making adding a `failOnEmptyInput` config property even more complicated, as the culprit in this case was of Gradle and of how task execution is computed.

> IMO a static analysis tool which tries to detect problems should also detect a wrong setup which is a problem as well and IMO a big problem because it covers up all real problems sitting there in the code.

Agree. Though in this specific case you instructed Detekt to pick an input that turned out being empty. I have several examples of Gradle modules that have no source code but have Detekt applied (a BOM, an Android resource only module, etc.). For those modules is totally reasonable to have Detekt just being skipped and resulting in a success.
The problem was that what you consider a "wrong setup" could be valid instead for another use case.

What we could do is list the snippet I posted in our official documentation, so others can benefit from it.

**Comment by robstoll** [comment]
Thanks for the analysis. Surely good to include the snippet 🙂👍
Personally, I would add a flag doNotFailOnEmptyInput which one needs to use in case of a BOM pom project or similar. I would even go that far to not provide a flag at all. Instead, such projects should simply not apply detekt because its for nothing. Or does detekt also check non-source related stuff?

**Comment by cortinico** [comment]
> Instead, such projects should simply not apply detekt because its for nothing

Agree, but if you use `subprojects {}` or `allprojects {}` block in your top level `build.gradle` file (as a lot of our users are doing), you're applying the plugin to all the modules. 
The current behavior makes sure those modules with not source are not breaking your overall builds.

**Comment by robstoll** [comment]
I most of the time use subprojects or configure(...) instead of a build.gradle in the subproject. So I would do the following
```
configure(subprojects.filter{ !it.name.contains("-bom") }){
  apply(...)
}
//or
subprojects {
  if(!it.name.contains("-bom")) apply(...)
}
```
And there you have your flag. No big deal IMO. But I see that you are hesitant to take a more restrictive approach than the current one in the sense of fail-if-no-input by default. Fine with me, I brought up my points, I think it's clear by now that both approaches require more or less the same amount of implementation in detekt and for the workaround. In the end, members of detekt need to decide more on a principal level IMO.

**Comment by BraisGabin** [comment]
I agree with @schalkms and @cortinico. Your point have sense but it could break other users flows. It seems that the use cases where you want empty source sets are legit. So I'm going to close this issue.

I do appreciate this kind of uses related with UX. Few people report things like this. But I think that in this case it's better to keep the plugin as it is now.
